### PR TITLE
ci(headless,atomic): fixed build version in prod

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/KIT-551

The build version in prod was wrong because the version uploaded to prod was not the newly bumped version.

To fix the issue, I changed the workflow in the deployment pipeline to "gitflow", which allows the "qa" and "prd" environments to run only on release (non-alpha) bump commits.

# Steps by environment before VS after

## Before

|#|dev (on alpha bump)        |qa (on dev approval)        |prd (on QA approval)          |
|-|---------------------------|----------------------------|------------------------------|
|1|Run tests                  |Upload alpha builds in qa S3|Upload alpha builds to prod S3|
|2|Publish alpha builds to NPM|Create release bump commit  |Publish release builds to NPM |
|3|Upload dev builds to S3    |Publish beta builds to NPM  |Notify doc repo of change     |
|4|[Wait for dev approval]    |[Wait for QA approval]      |                              |
|5|Start qa environment       |Start prd environment       |                              |

## After

|#|dev (on alpha bump)          |qa (on release bump)          |prd (on QA approval)            |
|-|-----------------------------|------------------------------|--------------------------------|
|1|Run tests                    |Run tests                     |Upload release builds to prod S3|
|2|Publish alpha builds to NPM  |Upload release builds to qa S3|Publish release builds to NPM   |
|3|Upload alpha builds to dev S3|[Wait for QA approval]        |Notify doc repo of change       |
|4|[Wait for manual approval]   |Start prd environment         |                                |
|5|Create release bump commit   |                              |                                |
|6|Publish beta builds to NPM   |                              |                                |

### CODE REVIEW

File: .deployment.config.json
Comment (btaillon) [inline .deployment.config.json:4]: Internal doc:
* https://tools-docs.dep.cloud.coveo.com/deployment-package/_static/schema_doc.html#development_workflow
* https://coveord.atlassian.net/wiki/spaces/CM/pages/1807712653/Following+up+with+a+deployment#For-gitflow-applications

File: .deployment.config.json
Comment (btaillon) [inline .deployment.config.json:13]: We don't want `dev` and `qa` to require manual approval in order to start their phases.

`qa` is now started automatically when there's a release bump commit.

File: .deployment.config.json
Comment (btaillon) [inline .deployment.config.json:19]: This ensures the `prd` environment will only be executed when there's manual approval from QA.

File: .deployment.config.json
Comment (btaillon) [inline .deployment.config.json:48]: `VERSION` is now the full version. We don't need a `PRERELEASE` variable anymore.

File: .deployment.config.json
Comment (btaillon) [inline .deployment.config.json:84]: It's a little weird, but now the `JenkinsfileQARelease` file is executed in the `dev` branch. I made do this since `JenkinsfileQARelease` is responsible for creating a release bump commit. The above "manual" step replaces the old dev approval.

File: .deployment.config.json
Comment (btaillon) [inline .deployment.config.json:68]: This step waits for approval from the dev team. This is how we'll trigger new releases now. QA approval will still be necessary for prod though.

File: Jenkinsfile
Comment (btaillon) [inline Jenkinsfile:102]: In a `gitflow` workflow, this `target-environment` parameter allows `dev` to be skipped altogether. This is necessary to start the `dev` pipeline exclusively on alpha bump commits and the `qa`+`prd` pipelines exclusively on release bump commits.

File: Jenkinsfile
Comment (btaillon) [inline Jenkinsfile:107]: It's not necessary to split everything across multiple lines, but I found it easier to read and review.

File: .deployment.config.json
Comment (dblanchette) [reply .deployment.config.json:19]: This is ok, but you could also do the opposite and have the default be false and set to true for dev and qa. That way, if you ever decide to add another environment (e.g.: hip), it won't start automatically.

**Review by dblanchette — APPROVED** [review]
LGTM

File: .deployment.config.json
Comment (samisayegh) [reply .deployment.config.json:68]: Why is this manual step needed?

File: Jenkinsfile
Comment (samisayegh) [reply Jenkinsfile:102]: I see, so I think the problem is on going into QA, there is a version bump to graduate the release.

However, with two deployment pipeline packages, this means one would approve to go to QA on one pipeline, but then release to prod using a second pipeline? Would that be correct?

If so, I feel it could be confusing, and would prefer if we could do it in one. I think the problem is the graduation version bump, which changes the commit to release?

Would it simplify things if we removed `-alpha` versions all together, and just followed semantic versioning? These versions would initially be tagged on npm as `alpha`. The idea being if a specific version is prod-ready, then we can just change it's npm tag, rather than needing to graduate it's version in a separate commit and start a new pipeline.

File: .deployment.config.json
Comment (btaillon) [reply .deployment.config.json:4]: my bad, I meant that the `dev` pipeline is run on a alpha bump commit in the first place, I'll correct this. Alpha bumps are done prior to starting the pipeline as usual.

File: .deployment.config.json
Comment (btaillon) [reply .deployment.config.json:68]: I removed the manual approval (Deploy button in Slack) that was required to transition to QA in favor of adding one here instead so that we can create a release bump commit in the dev pipeline after approval.

File: Jenkinsfile
Comment (btaillon) [reply Jenkinsfile:102]: Essentially yes, approving an alpha bump creates a release bump which goes directly to the qa pipeline and requires QA approval before finally moving on to prod.

If we remove `-alpha`, we can go back to a single deployment package. The only negative then would be that versions read inside the build wouldn't have any indicator of whether they're alpha or not.

**Review by ThibodeauJF — APPROVED** [review]

**Review by nbuissoncoveo — APPROVED** [review]

File: Jenkinsfile
Comment (nbuissoncoveo) [reply Jenkinsfile:102]: What exactly is an 'alpha' bump?

Comment (samisayegh) [reply Jenkinsfile:102]: An alpha bump is a version bump where the version ends with `-alpha`.

Regarding the indicator, I wonder how useful it is because most people will never install these packages. For example, if

1.0.0 is latest

then we publish 1.1.0 with an alpha tag
then we publish 1.2.0 with a latest tag

I expect someone upgrading will go straight to 1.2.0,

Generally, I think all versions are relatively stable, since even those released on merge into master must pass FTs.
With that backdrop, I was wondering if we can make the release process simpler since I think jumping from one pipeline to the next is not dev friendly,

**Pull request closed without merging**
